test/distributed/_composable/test_composability/test_2d_composability.py

```diff
@@ -2,30 +2,136 @@
 from typing import List, Type
 
 import torch
+
+import torch.distributed as dist
 import torch.distributed.checkpoint as dcp
 import torch.nn as nn
+
+import torch.nn.functional as F
 from torch.distributed._composable import replicate
 from torch.distributed._composable.fsdp import CPUOffloadPolicy, fully_shard
-from torch.distributed._tensor import init_device_mesh
+from torch.distributed._tensor import (
+    DTensor,
+    DTensor as DT,
+    init_device_mesh,
+    Replicate,
+    Shard,
+)
 from torch.distributed._tensor.debug.comm_mode import CommDebugMode
 from torch.distributed.checkpoint.state_dict import (
     get_model_state_dict,
     get_optimizer_state_dict,
+    set_optimizer_state_dict,
 )
 from torch.distributed.device_mesh import DeviceMesh
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.distributed.fsdp._common_utils import (
+    _get_module_fsdp_state,
+    clean_tensor_name,
+)
+from torch.distributed.fsdp.fully_sharded_data_parallel import StateDictType
+from torch.distributed.tensor.parallel import (
+    ColwiseParallel,
+    parallelize_module,
+    RowwiseParallel,
+)
+from torch.distributed.tensor.parallel.ddp import _pre_dp_module_transform
+from torch.distributed.tensor.parallel.fsdp import DTensorExtensions
+from torch.distributed.tensor.parallel.input_reshard import input_reshard
+
+from torch.nn.parallel import DistributedDataParallel as DDP
 from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
 from torch.testing._internal.common_fsdp import FSDPTest, MLP, MLPStack
-from torch.testing._internal.common_utils import run_tests, skipIfRocm
+
+from torch.testing._internal.common_utils import (
+    instantiate_parametrized_tests,
+    parametrize,
+    run_tests,
+    skipIfRocm,
+)
+
 from torch.testing._internal.distributed._tensor.common_dtensor import (
+    DTensorTestBase,
+    MLPModule,
     ModelArgs,
     Transformer,
+    with_comms,
 )
 from torch.testing._internal.distributed.checkpoint_utils import with_temp_dir
 
 
+def init_model(device_type, model_parallel_size=2):
+    torch.manual_seed(0)
+    model = MLPModule(device_type)
+    torch.manual_seed(0)
+    twod_model = MLPModule(device_type)
+    model = DDP(model)
+
+    # 2-D mesh is [dp, tp]
+    world_size = dist.get_world_size()
+    mesh_2d = init_device_mesh(
+        device_type,
+        (world_size // model_parallel_size, model_parallel_size),
+        mesh_dim_names=("dp", "tp"),
+    )
+
+    dp_pg = mesh_2d.get_group(mesh_dim=0)
+
+    parallelize_plan = {
+        "net1": ColwiseParallel(),
+        "net2": RowwiseParallel(),
+    }
+    twod_model = parallelize_module(twod_model, mesh_2d["tp"], parallelize_plan)
+    _pre_dp_module_transform(twod_model)
+    # TODO: Add tests when using gradient_as_bucket_view and static_graph for DDP.
+    twod_model = DDP(twod_model, process_group=dp_pg)
+    return model, twod_model, dp_pg
+
+
+class SimpleModel(nn.Module):
```

Comment:
Not need to do in this PR, but we had so many duplicate codes in the past, it would be great if we can refactor them to consolidate them a little bit.